extensions/amp-react-img/0.1/amp-react-img.js

```diff
@@ -0,0 +1,165 @@
+    const key = keys[i];
+    const value = props[key];
+    if (value !== undefined) {
+      out[key] = value;
+    }
+  }
+
+  return out;
+}
+
+const ATTRIBUTES_TO_PROPAGATE = [
+  'alt',
+  'title',
+  'referrerpolicy',
+  'aria-label',
+  'aria-describedby',
+  'aria-labelledby',
+  'srcset',
+  'src',
+  'sizes',
+];
+
+AMP.extension('amp-react-img', '0.1', AMP => {
+  const react = requireExternal('react');
+  const {render} = requireExternal('react-dom');
+
+  /**
+   * We'll implement all our new extensions as React/Preact Components (TBD).
+   * They're true Components, not AmpElements/Amp.BaseElements.
+   */
+  class AmpImg extends react.Component {
+    /**
+     * @param {!Object} props
+     */
+    constructor(props) {
+      super(props);
+
+      this.state = {prerender: true};
+    }
+
+    /**
+     * @return {*}
+     */
+    render() {
+      const props = pick(this.props, ATTRIBUTES_TO_PROPAGATE);
+      const {id, 'i-amphtml-ssr': ssr} = this.props;
+      if (ssr) {
+        // TODO: Figure out SSR children
+      }
+
+      if (id) {
+        props['amp-img-id'] = id;
+      }
+      props['decoding'] = 'async';
+
+      // amp-img is always allowed to render, but we want to make this interesting...
+      const {prerender} = this.state;
+      if (prerender) {
+        delete props['src'];
+        delete props['srcset'];
+      }
+
+      return react.createElement('img', props);
+    }
+  }
+
+  /**
+   * ReactCompatibleBaseElement is a compatibility wrapper around AMP's
+   * BaseElement. It takes a Component to compose, and calls renders the
+   * component with any state necessary.
+   *
+   * This code can be shared across multiple extensions, all they need to do is
+   * supply the Component.
+   *
+   * This is only necessary in when in AMP pages. If we're in a Bento page,
+   * this code is useless. We'll supply a CustomElement factory class that will
+   * provide compatibilty between CE and React.
+   *
+   * @param {!React.Component} Component
+   * @return {Amp.BaseElement}
+   */
+  function ReactCompatibleBaseElement(Component) {
+    return class ReactBaseElement extends AMP.BaseElement {
+      /** @param {!AmpElement} element */
+      constructor(element) {
+        super(element);
+
+        /** @private {?Component} */
+        this.el_ = null;
+      }
+
+      /**
+       * For demo purposes.
+       * @override
+       */
+      renderOutsideViewport() {
+        return false;
+      }
+
+      /** @override */
+      isLayoutSupported() {
+        return true;
+      }
+
+      /** @override */
+      buildCallback() {
+        const el = react.createElement(Component, this.attributes_());
+        this.el_ = render(el, this.element);
+      }
+
+      /** @override */
+      layoutCallback() {
+        const el = devAssert(this.el_);
+        el.setState({prerender: false});
+
+        const rerender = react.createElement(Component, this.attributes_());
```

Comment:
Shouldn't we only create the element once and then push new state into it?

Reply:
Despite the name, this won't actually recreate the element. It creates a JSX node, something like `{ tagName: Component, props: { src: 'foo.com' }, children: [] }`. The `render` call below will reconcile that JSX node with the existing JSX node tied to our original render.

I actually tried only setting the state on the `this.el_`, but it wouldn't update to render the `src`. It was because I didn't call `render` with the new JSX node to cause the diffing.

Reply:
Hmm. The call to `render()` creates a new JSX node. But why do we need multiple instances of `AmpImg` component? What's the cost of that?

Reply:
It's a reference to the constructor function, not an instance of one. React creates the instance (if necessary) during diffing. Because there's already an instance attached to this element, it'll reuse it.

Reply:
Hmm. That's cool.